packages/react-next/src/components/Fabric/Fabric.base.tsx

```diff
@@ -0,0 +1,94 @@
+import { IProcessedStyleSet } from '@uifabric/merge-styles';
+import { ITheme, createTheme } from '../../Styling';
+import { useMergedRefs } from '@uifabric/react-hooks';
+import { useFocusRects } from '@uifabric/utilities';
+
+const getClassNames = classNamesFunction<IFabricStyleProps, IFabricStyles>();
+const getFabricTheme = memoizeFunction((theme?: ITheme, isRTL?: boolean) => createTheme({ ...theme, rtl: isRTL }));
+
+const getDir = ({ theme, dir }: IFabricProps) => {
+  const contextDir = getRTL(theme) ? 'rtl' : 'ltr';
+  const pageDir = getRTL() ? 'rtl' : 'ltr';
+  const componentDir = dir ? dir : contextDir;
+  return {
+    // If Fabric dir !== contextDir
+    // Or If contextDir !== pageDir
+    // Then we need to set dir of the Fabric root
+    rootDir: componentDir !== contextDir || componentDir !== pageDir ? componentDir : dir,
+    // If dir !== contextDir || pageDir
+    // then set contextual theme around content
+    needsTheme: componentDir !== contextDir,
+  };
+};
+
+// tslint:disable-next-line:function-name no-function-expression
+export const FabricBase = React.forwardRef(function(props: IFabricProps, ref: React.Ref<HTMLDivElement>) {
+  const { className, theme, applyTheme, applyThemeToBody } = props;
+
+  const classNames = getClassNames(getStyles, {
+    theme: theme!,
+    applyTheme: applyTheme,
+    className,
+  });
+
+  const rootElement = React.useRef<HTMLDivElement | null>(null);
+  useApplyThemeToBody(applyThemeToBody, classNames, rootElement);
+  useFocusRects(rootElement);
+
+  return <>{useRenderedContent(props, classNames, rootElement, ref)}</>;
+});
+
+function useRenderedContent(
+  props: IFabricProps,
+  { root }: IProcessedStyleSet<IFabricStyles>,
+  rootElement: React.RefObject<HTMLDivElement | undefined>,
+  ref: React.Ref<HTMLDivElement>,
+) {
+  const { as: Root = 'div', dir, theme } = props;
+  const divProps = getNativeProps<React.HTMLAttributes<HTMLDivElement>>(props, divProperties, ['dir']);
+
+  const { rootDir, needsTheme } = getDir(props);
+
+  let renderedContent = <Root dir={rootDir} {...divProps} className={root} ref={useMergedRefs(rootElement, ref)} />;
+
+  // Create the contextual theme if component direction does not match parent direction.
+  if (needsTheme) {
+    renderedContent = (
+      <Customizer settings={{ theme: getFabricTheme(theme, dir === 'rtl') }}>{renderedContent}</Customizer>
+    );
+  }
+
+  return renderedContent;
+}
+
+function useApplyThemeToBody(
+  applyThemeToBody: boolean | undefined,
+  { bodyThemed }: IProcessedStyleSet<IFabricStyles>,
+  rootElement: React.RefObject<HTMLDivElement | undefined>,
+) {
+  React.useEffect((): void | (() => void) => {
+    if (applyThemeToBody) {
+      const currentDoc = getDocument(rootElement.current);
+      if (currentDoc) {
+        currentDoc.body.classList.add(bodyThemed);
+        return () => {
+          currentDoc.body.classList.remove(bodyThemed);
+        };
+      }
+    }
+  }, [bodyThemed]);
```

Comment:
i think we need `rootElement.current` here too

Reply:
A change to the ref value doesn't take effect until the render is pushed to the DOM, won't trigger a rerender. If we're concerned about it, we'd have to use a state value instead.

Reply:
makes sense